llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -694,6 +698,29 @@ void AIEPostRASchedStrategy::handleRegionConflicts(
   }
 }
 
+/// The earliest use of this instruction in the next iteration.
+/// Note that we reason with "bottom-up" cycle, so a larger cycle means it's
+/// used earlier in topological order. If the SU has no loop-carried dependency,
+/// this will be MAX_INT.
+int getEarliestLoopCarriedUse(const SUnit &SU,
+                              const InterBlockEdges &LoopEdges) {
+  const SUnit *SUInCurrIter = LoopEdges.getPreBoundaryNode(SU.getInstr());
+  assert(SUInCurrIter);
+  assert(SUInCurrIter->getHeight() >= SU.getHeight());
+
+  // Look at loop-carried dependencies to see how early the instruction will be
+  // needed in the next iteration.
+  int EarliestCycle = std::numeric_limits<int>::max();
+  for (const SDep &Succ : SUInCurrIter->Succs) {
+    if (!LoopEdges.isPostBoundaryNode(Succ.getSUnit()))
+      continue;
+
+    EarliestCycle = std::min(EarliestCycle, int(Succ.getSUnit()->getHeight()));
```

Comment:
Yes. BotReadyCycle is in Height currency, TopReadyCycle is in Depth currency. Let's introduce strong integer types for them.